modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/PartitionCommandListener.java

```diff
@@ -0,0 +1,166 @@
+import org.apache.ignite.raft.client.WriteCommand;
+import org.apache.ignite.raft.client.service.CommandClosure;
+import org.apache.ignite.raft.client.service.RaftGroupCommandListener;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Partition command handler.
+ */
+public class PartitionCommandListener implements RaftGroupCommandListener {
+    /** Storage. */
+    private HashMap<KeyWrapper, BinaryRow> storage = new HashMap<>();
+
+    /** {@inheritDoc} */
+    @Override public void onRead(Iterator<CommandClosure<ReadCommand>> iterator) {
+        while (iterator.hasNext()) {
+            CommandClosure<ReadCommand> clo = iterator.next();
+
+            assert clo.command() instanceof GetCommand;
+
+            clo.success(new KVGetResponse(storage.get(extractAndWrapKey(((GetCommand)clo.command()).getKeyRow()))));
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onWrite(Iterator<CommandClosure<WriteCommand>> iterator) {
+        while (iterator.hasNext()) {
+            CommandClosure<WriteCommand> clo = iterator.next();
+
+            if (clo.command() instanceof InsertCommand) {
+                BinaryRow previous = storage.putIfAbsent(
+                    extractAndWrapKey(((InsertCommand)clo.command()).getRow()),
+                    ((InsertCommand)clo.command()).getRow()
+                );
+
+                clo.success(previous == null);
+            }
+            else if (clo.command() instanceof DeleteCommand) {
+                BinaryRow deleted = storage.remove(
+                    extractAndWrapKey(((DeleteCommand)clo.command()).getKeyRow())
+                );
+
+                clo.success(deleted != null);
+            }
+            else if (clo.command() instanceof ReplaceCommand) {
+                ReplaceCommand cmd = ((ReplaceCommand)clo.command());
+
+                BinaryRow expected = cmd.getOldRow();
+
+                KeyWrapper key = extractAndWrapKey(expected);
+
+                BinaryRow current = storage.get(key);
+
+                if ((current == null && !expected.hasValue()) ||
+                    equalValues(current, expected)) {
+                    storage.put(key, cmd.getRow());
+
+                    clo.success(true);
+                }
+                else
+                    clo.success(false);
+            }
+            else if (clo.command() instanceof UpsertCommand) {
+                storage.put(
+                    extractAndWrapKey(((UpsertCommand)clo.command()).getRow()),
+                    ((UpsertCommand)clo.command()).getRow()
+                );
+
+                clo.success(null);
+            }
+            else
+                assert false : "Command was not found [cmd=" + clo.command() + ']';
+        }
+    }
+
+    /**
+     * @param row Row.
+     * @return Extracted key.
+     */
+    @NotNull private boolean equalValues(@NotNull BinaryRow row, @NotNull BinaryRow row2) {
+        if (row.hasValue() ^ row2.hasValue())
```

Comment:
Doesn't look correct.  If both rows have values it will return false.

Reply:
The `^` operator returns true when a left part is different from a right part.
false ^ false = false
false ^ true = true
true ^ false = true
true ^ true = false

Reply:
This `^` operator returns true when the left part is different to the right one.
false ^ false = false
false ^ true = true
true ^ false = true
true ^ true = false